File: client/tests/test_framer.py
"""Unit tests for the LengthPrefixedFramer adapter."""

import struct

import pytest

from src.adapters.length_prefixed_framer import LengthPrefixedFramer


@pytest.fixture
def framer() -> LengthPrefixedFramer:
    return LengthPrefixedFramer()

def test_pack_unpack_symmetry_small(framer: LengthPrefixedFramer) -> None:
    payload = b"Hello"
    assert framer.unpack(framer.pack(payload)) == payload


def test_pack_unpack_symmetry_single_byte(framer: LengthPrefixedFramer) -> None:
    payload = b"\xff"
    assert framer.unpack(framer.pack(payload)) == payload


def test_pack_unpack_symmetry_medium(framer: LengthPrefixedFramer) -> None:
    payload = bytes(range(256))
    assert framer.unpack(framer.pack(payload)) == payload


# ---------- empty payload ----------


def test_empty_payload(framer: LengthPrefixedFramer) -> None:
    payload = b""
    packed = framer.pack(payload)

    # Frame should be exactly 4 bytes: 00 00 00 00
    assert len(packed) == 4
    assert packed == b"\x00\x00\x00\x00"

    assert framer.unpack(packed) == b""


# ---------- large payload ----------


def test_large_payload(framer: LengthPrefixedFramer) -> None:
    payload = b"\xab" * 100_000
    packed = framer.pack(payload)
    assert len(packed) == 4 + len(payload)
    assert framer.unpack(packed) == payload


# ---------- header encoding ----------


def test_header_is_big_endian(framer: LengthPrefixedFramer) -> None:
    payload = b"\x00" * 256
    packed = framer.pack(payload)
    assert packed[:4] == b"\x00\x00\x01\x00"


# ---------- error cases ----------


def test_unpack_too_short_for_header(framer: LengthPrefixedFramer) -> None:
    with pytest.raises(ValueError, match="too short for header"):
        framer.unpack(b"\x00\x00")


def test_unpack_too_short_for_payload(framer: LengthPrefixedFramer) -> None:
    stream = struct.pack("!I", 10) + b"\x01\x02"
    with pytest.raises(ValueError, match="too short for payload"):
        framer.unpack(stream)

File: server/tests/test_framer.cpp
#include <cstdint>
#include <memory>
#include <numeric>
#include <vector>

#include <gtest/gtest.h>

#include "adapters/length_prefixed_framer.h"
#include "ports/i_framer.h"

using iris::IFramer;
using iris::LengthPrefixedFramer;

class FramerTest : public ::testing::Test {
protected:
    std::unique_ptr<IFramer> framer = std::make_unique<LengthPrefixedFramer>();
};

TEST_F(FramerTest, PackUnpackSymmetry_SmallPayload) {
    std::vector<uint8_t> payload = {0x48, 0x65, 0x6C, 0x6C, 0x6F};  // "Hello"
    auto packed = framer->pack(payload);
    auto unpacked = framer->unpack(packed);
    EXPECT_EQ(unpacked, payload);
}

TEST_F(FramerTest, PackUnpackSymmetry_SingleByte) {
    std::vector<uint8_t> payload = {0xFF};
    auto packed = framer->pack(payload);
    auto unpacked = framer->unpack(packed);
    EXPECT_EQ(unpacked, payload);
}

TEST_F(FramerTest, PackUnpackSymmetry_MediumPayload) {
    std::vector<uint8_t> payload(256);
    std::iota(payload.begin(), payload.end(), 0);  // 0x00..0xFF
    auto packed = framer->pack(payload);
    auto unpacked = framer->unpack(packed);
    EXPECT_EQ(unpacked, payload);
}

// ---------- empty payload ----------

TEST_F(FramerTest, EmptyPayload) {
    std::vector<uint8_t> payload;
    auto packed = framer->pack(payload);

    ASSERT_EQ(packed.size(), 4u);
    EXPECT_EQ(packed[0], 0x00);
    EXPECT_EQ(packed[1], 0x00);
    EXPECT_EQ(packed[2], 0x00);
    EXPECT_EQ(packed[3], 0x00);

    auto unpacked = framer->unpack(packed);
    EXPECT_TRUE(unpacked.empty());
}

// ---------- large payload ----------

TEST_F(FramerTest, LargePayload) {
    std::vector<uint8_t> payload(100'000, 0xAB);
    auto packed = framer->pack(payload);
    EXPECT_EQ(packed.size(), 4 + payload.size());

    auto unpacked = framer->unpack(packed);
    EXPECT_EQ(unpacked, payload);
}

// ---------- header encoding ----------

TEST_F(FramerTest, HeaderIsBigEndian) {
    std::vector<uint8_t> payload(256, 0x00);
    auto packed = framer->pack(payload);

    EXPECT_EQ(packed[0], 0x00);
    EXPECT_EQ(packed[1], 0x00);
    EXPECT_EQ(packed[2], 0x01);
    EXPECT_EQ(packed[3], 0x00);
}

// ---------- error cases ----------

TEST_F(FramerTest, UnpackTooShortForHeader) {
    std::vector<uint8_t> stream = {0x00, 0x00};
    EXPECT_THROW(framer->unpack(stream), std::runtime_error);
}

TEST_F(FramerTest, UnpackTooShortForPayload) {
    std::vector<uint8_t> stream = {0x00, 0x00, 0x00, 0x0A, 0x01, 0x02};
    EXPECT_THROW(framer->unpack(stream), std::runtime_error);
}
